R/zchunk_L100.GDP_hist.R

```diff
@@ -34,13 +34,46 @@ module_socioeconomics_L100.GDP_hist <- function(command, ...) {
     usda_gdp_mer <- get_data(all_data, "socioeconomics/USDA_GDP_MER")
     assert_that(tibble::is.tibble(usda_gdp_mer))
 
-    # Convert to long form, filter to historical years, convert units
+    # Convert to long form, convert units
     usda_gdp_mer %>%
       select(-Country) %>%
       gather_years %>%
       filter(!is.na(value), !is.na(iso)) %>%
       mutate(value = value * CONV_BIL_MIL * gdp_deflator(1990, base_year = 2010),
-             year = as.integer(year)) %>%
+             year = as.integer(year)) ->
+      long_iso_year_gdp
+
+
+    # Perform BYU
+    # Skeleton - constant extrapolation
+    # BYUcompliant
+    #
+    # BYU NOTE - must make sure the units are all 1990 USD (or at least the
+    # same year basis) so averaging behaves. Fine here, but worth noting
+    # across other chunk where money comes up, especially since we drop
+    # unit information pretty early on in most chunks.
+    # BYU NOTE - need to think about labeling that the output has BYU
+    # update done, and a note about the method?
+    if(max(long_iso_year_gdp$year) < BYU_YEAR){
+
+      missingyears <- (max(long_iso_year_gdp$year) +  1):BYU_YEAR
+
+      # Constant extrapolation operating only on
+      # numeric vector. Can operate on any numeric vector, regardless of grouping
+      # or column name.
+      # What gets output in this chunk
+      long_iso_year_gdp %>%
+        complete(year = c(year, missingyears), iso) %>%
+        group_by(iso) %>%
+        mutate(value = extrapolate_constant(value, n = 1,
+                                            numMissing = length(missingyears))) %>%
```

Comment:
In this example calculating `numMIssing` is straightforward.  But I think the typical use case will be that users just want to: `complete(year = HISTORICAL_YEARS, non, year, columns)` and may not explicitly know what years were originally present in their data set.

They could define some constant about what they are, and we do have many of these constants, but that is going to proliferate constants and the constant would need to get updated anytime the data set does.

Users could just figure it out by doing something like:
```
long_iso_year_gdp %>%
	        group_by(iso) %>%
                mutate(numMissing = BYU_YEAR - max(year)) %>% # BYU_YEAR for now, max(HISTORICAL_YEARS) at some point
	        complete(year = c(year, missingyears), iso) %>%

	        mutate(value = extrapolate_constant(value, n = 1, numMissing = numMissing) %>% # we get the whole vector but they are all the same
```
Which seems a bit cumbersome / wasteful.

What about in `extrapolate_constant` just filling from the "last non-NA value + 1" to the end?  That way it can be self-contained?

Reply:
>What about in extrapolate_constant just filling from the "last non-NA value + 1" to the end? That way it can be self-contained?

That was my thought too, on reading through the `extrapolate_constant()` code. Calculating `numMissing` seems like a huge drag.

The potential problem (recapping some of previous discussion) is that if there's a time series with final NA, this will overwrite that NA?

Reply:
Also, right now the input argument `n` to `extrapolate_cosntate()`specifies how many of the last non-NA values to average to get a value for filling in tail end NA's. The idea is obviously that everyone will just use `n=1`, projecting the last reported value forward. 
However, if a value `n>1` is input, filling in the last "last non-NA value + 1" will overwrite a reported value. 
I think the case where 2010 value = NA so we take 2009 value and fill it in for 2011-onward but leave 2010 as NA is definitely weird and we would want to overwrite that final NA? One (possibly difficult) option would be to actually see if this comes up anywhere in the data system and if there's any special way of handling it other than 2010value=2009value when it does.
If we're ok with overwriting the final NA value by doing the last missingYears+1, then I do think we need to drop support for arbitrary `n` values as input and force that value to just be 1